lib/libzpool/uzfs_rebuilding.c

```diff
@@ -0,0 +1,296 @@
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+#include <sys/dmu_objset.h>
+#include <sys/uzfs_zvol.h>
+#include <sys/dmu_traverse.h>
+#include <sys/dsl_pool.h>
+#include <sys/dsl_dataset.h>
+#include <sys/dsl_destroy.h>
+#include <sys/dmu_tx.h>
+#include <uzfs_io.h>
+#include <uzfs_rebuilding.h>
+
+#define	ADD_TO_IO_CHUNK_LIST(list, e_offset, e_len, count)		\
+	do {    							\
+		uzfs_io_chunk_list_t  *node;				\
+		node = umem_alloc(sizeof (*node), UMEM_NOFAIL);         \
+		node->offset = e_offset;                                \
+		node->len = e_len;                                      \
+		list_insert_tail(list, node);                           \
+		count++;                                                \
+	} while (0)
+
+int
+compare_blk_metadata(blk_metadata_t *first, blk_metadata_t *second)
+{
+	if (first->io_num < second->io_num)
+		return (-1);
+	if (first->io_num == second->io_num)
+		return (0);
+	return (1);
+}
+
+boolean_t
+iszero(blk_metadata_t *md)
+{
+	if (md->io_num == 0)
+		return (B_TRUE);
+	return (B_FALSE);
+}
+
+#define	EXECUTE_DIFF_CALLBACK(last_lun_offset, diff_count, buf, 	\
+    last_index, arg, last_md, zv, func)					\
+		do {							\
+			func(last_lun_offset, diff_count * 		\
+			    zv->zv_metavolblocksize, (blk_metadata_t *) \
+			    (buf + last_index), zv->zv_objset, arg);	\
+			diff_count = 0;					\
+			last_index = 0;					\
+			last_md = NULL;					\
+			diff_count = 0;					\
+		} while (0)
+
+int
+get_metadata_snapshot_info(zvol_state_t *zv, blk_metadata_t *md,
+    zvol_state_t **snap_zv)
+{
+	char *snap_name, *dataset;
+	int ret = 0;
+	zvol_state_t *s_zv;
+	objset_t *s_obj;
+
+	dataset = kmem_asprintf("%s@%s%lu", zv->zv_name,
+	    IO_DIFF_SNAPNAME, md->io_num);
+
+	ret = dmu_objset_own(dataset, DMU_OST_ANY, B_TRUE, snap_zv, &s_obj);
+	if (ret == ENOENT) {
+		snap_name = kmem_asprintf("%s%llu", IO_DIFF_SNAPNAME,
+		    md->io_num);
+
+		ret = dmu_objset_snapshot_one(zv->zv_name, snap_name);
+		if (ret) {
+			printf("Failed to create snapshot for %s\n",
+			    zv->zv_name);
+			strfree(dataset);
+			strfree(snap_name);
+			return (ret);
+		}
+
+		strfree(snap_name);
+		ret = dmu_objset_own(dataset, DMU_OST_ANY, B_TRUE, snap_zv,
```

Comment:
how about using uzfs_open_dataset_init, which creates and initializes zv..?

Reply:
yes. we can use it. I will try that and update changes.